### PR TITLE
Add a default Dockerfile if not exists for java project

### PR DESCRIPTION
**Pull request by haoozhang**
Add a default Dockerfile if not exists for a java project.

### CODE REVIEW

File: cli/azd/internal/appdetect/docker.go [hunk 1]
```diff
@@ -70,3 +82,50 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+	file, err := os.Create(dockerfilePath)
+	if err != nil {
+		return fmt.Errorf("failed to create Dockerfile at %s: %w", dockerfilePath, err)
+	}
+	defer file.Close()
+
+	// for single-module project, we have to maven build first, then copy and run jar
+	// for multi-module project, just copy and run jar because of prepackage hook
+	var dockerfileContent string
+	if hasParentPom {
```
Comment (haoozhang) [inline cli/azd/internal/appdetect/docker.go:108]: Here I don't use the `draft` to generate Dockerfile, because its template seems not meet our need, for example, in the [draft Docker-java template](https://github.com/Azure/draft/blob/main/template/dockerfiles/java/Dockerfile), it always try to find the `-SNAPSHOT.jar`, but we should not ensure this.

File: cli/azd/internal/appdetect/docker.go [hunk 2]
```diff
@@ -70,3 +82,50 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+		return fmt.Errorf("error checking Dockerfile at path %s: %w", path, err)
+	}
+
+	file, err := os.Create(dockerfilePath)
+	if err != nil {
+		return fmt.Errorf("failed to create Dockerfile at %s: %w", dockerfilePath, err)
+	}
+	defer file.Close()
+
+	// for single-module project, we have to maven build first, then copy and run jar
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:105]: nit: `maven build` -> `run mvn package`

File: cli/azd/internal/appdetect/docker.go [hunk 3]
```diff
@@ -70,3 +82,50 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+	// for multi-module project, just copy and run jar because of prepackage hook
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:106]: `because of prepackage hook` -> `because 'mvn package' already executed in prepackage hook`.

File: cli/azd/internal/appdetect/docker.go [hunk 4]
```diff
@@ -11,14 +11,26 @@ import (
...
 	for _, entry := range entries {
 		if strings.ToLower(entry.Name()) == "dockerfile" {
 			dockerFilePath := filepath.Join(path, entry.Name())
 			return AnalyzeDocker(dockerFilePath)
 		}
 	}
 
+	// if Dockerfile not exists, provide a default one
+	if project.Language == Java {
+		_, hasParentPom := project.Options[JavaProjectOptionMavenParentPath]
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:25]: Is it possible that the parent is not accurate? For example: The root pom has some modules: module-a, module-b, but current module (module-c) is not set the root pom as parent. Will this case supported in our current code? Could you please add this test case in unit test?

File: cli/azd/internal/appdetect/docker.go [hunk 5]
```diff
@@ -70,3 +82,50 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+	defer file.Close()
+
+	// for single-module project, we have to maven build first, then copy and run jar
+	// for multi-module project, just copy and run jar because of prepackage hook
+	var dockerfileContent string
+	if hasParentPom {
+		dockerfileContent = `FROM mcr.microsoft.com/openjdk/jdk:17-distroless
+COPY ./target/*.jar app.jar
+EXPOSE 8080
+ENTRYPOINT ["java", "-jar", "/app.jar"]`
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:112]: nit: Move this string out of func, and define a const, do like this will make code look better.

File: cli/azd/internal/appdetect/docker.go [hunk 6]
```diff
@@ -70,3 +82,50 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+	} else {
+		dockerfileContent = `FROM maven:3 AS build
+WORKDIR /app
+COPY . .
+RUN mvn --batch-mode clean package -DskipTests
+
+FROM mcr.microsoft.com/openjdk/jdk:17-distroless
+WORKDIR /
+COPY --from=build /app/target/*.jar /app.jar
+EXPOSE 8080
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:122]: 1. If it's a JMS /  spring cloud stream app with out spring-boot-web dependency, expose 8080 may not the right way. 
2. If in customer's application.yaml, the port is set to another one like this `server.port=8181`.
3. Now for eureka server / config server, will they work well for current port setting?

File: cli/azd/internal/appdetect/docker.go [hunk 7]
```diff
@@ -11,14 +11,26 @@ import (
...
 		if strings.ToLower(entry.Name()) == "dockerfile" {
 			dockerFilePath := filepath.Join(path, entry.Name())
 			return AnalyzeDocker(dockerFilePath)
 		}
 	}
 
+	// if Dockerfile not exists, provide a default one
+	if project.Language == Java {
+		_, hasParentPom := project.Options[JavaProjectOptionMavenParentPath]
+		err := addDefaultDockerfile(path, hasParentPom)
```
Comment (rujche) [inline cli/azd/internal/appdetect/docker.go:26]: Is it necessary to delete the Dockerfile after used? It's valid that customer want to avoid the Dockerfile in his repo / folder. We shouldn't do much assumption in customer's repo / folder.

Comment (rujche) [reply cli/azd/internal/appdetect/docker.go:26]: Another problem is that not all java project will output a runnable jar.

File: cli/azd/internal/appdetect/docker.go (hunk 2 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:105]: Updated.

File: cli/azd/internal/appdetect/docker.go (hunk 3 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:106]: Updated.

File: cli/azd/internal/appdetect/docker.go (hunk 5 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:112]: Updated.

File: cli/azd/internal/appdetect/docker.go (hunk 4 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:25]: The answer is no, the current code does NOT support this scenario that the `module-c` is just in the sub folder but not set root pom as parent (or parent pom doesn't cover `module-c`). In this case, actually it's still a standalone project, but it will be mistakenly identified as a sub project from our code. It's because that when we detect parent pom, we currently detect parent pom based on just prefix matching (`we can say that the project is in the root project if the path is under the project`), which has this kind of limitation.
image

So, the key to this problem is to enhance the parent pom detection.

File: cli/azd/internal/appdetect/docker.go (hunk 6 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:122]: 1. I remove this line of `expose 8080` first, and then prompt user to input the desired port.
2. Currently we still not support detect the `server.port`. I think we need to enhance the port detection.

File: cli/azd/internal/appdetect/docker.go (hunk 7 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:26]: Updated the Dockerfile to support both jar and war.

Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:26]: > Is it necessary to delete the Dockerfile after used? It's valid that customer want to avoid the Dockerfile in his repo / folder. We shouldn't do much assumption in customer's repo / folder.

The Dockerfile generated during the `azd init` phase will be used when executing the `azd up` command (specifically during the `azd package` process). Even if it is deleted after the current round of `azd up`, subsequent executions of `azd up` will throw an exception because the Dockerfile is referenced in the `azure.yaml` file but is no longer exist.
So, it's difficult to determine when to delete it from our side. I personally think we can add a log message to notify users that the Dockerfile was auto-generated for them, if they don't use it anymore and don't want it, they can delete by themselves.

File: cli/azd/internal/appdetect/docker.go [hunk 8]
```diff
@@ -70,3 +82,62 @@ func parsePortsInLine(s string) ([]Port, error) {
 	}
 	return ports, nil
 }
+
+const (
+	DockerfileSingleStage = `FROM openjdk:17-jdk-slim
+COPY ./target/*.jar app.jar
```
Comment (saragluna) [inline cli/azd/internal/appdetect/docker.go:88]: what if users application requires a higher JDK version?

Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:88]: I intend to determine it by the `java.version` property, but it's not mandatory in spring boot project. Maybe we hardcode a higher version? or do you have any suggestion?

File: cli/azd/internal/appdetect/docker.go (hunk 4 above)
Comment (rujche) [reply cli/azd/internal/appdetect/docker.go:26]: > the key to this problem is to enhance the parent pom detection.

Could you please help to do this? Maybe in another PR after current PR merged.

File: cli/azd/internal/appdetect/docker.go (hunk 6 above)
Comment (rujche) [reply cli/azd/internal/appdetect/docker.go]: Could you please help to create an ADO item for this?

File: cli/azd/internal/appdetect/docker.go (hunk 7 above)
Comment (rujche) [reply cli/azd/internal/appdetect/docker.go:27]: > Updated the Dockerfile to support both jar and war.

I mean the jar file is not runnable, for example: one project has 3 modules: 
1. module-one-app
2. module-two-app
3. module-shared

The module-one-app and module-two-app will generate runnable jar. But module-shared will not output a runnable jar, it's just a library that be used in module-one-app and module-two-app.

File: cli/azd/internal/appdetect/docker.go (hunk 8 above)
Comment (rujche) [reply cli/azd/internal/appdetect/docker.go:89]: My suggestion:
1. Detect java version by these ways: 
  1.1. From pom files.
  1.2. From spring-boot version
  1.3. From "java" command in Path.
  1.4. Maybe other ways...
2. If can not be detected, use default version. (If Java version is back-ward compatible, use the higher version is better. For example: Java 21).

File: cli/azd/internal/appdetect/docker.go (hunk 4 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:26]: Sure, I have created a work item https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2330371, and will enhance it later.

File: cli/azd/internal/appdetect/docker.go (hunk 6 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: Yeah, I have created the item to track https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2330216, and will enhance it in another PR.

File: cli/azd/internal/appdetect/docker.go (hunk 7 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: azd itself is currently NOT aware of whether it's a runnable module or shared module (nor runnable). As long as a `pom.xml` detected, it will be identified as a java project. The current user experience is that customer can be able to confirm / add / remove the detected services. 
I put the adding Dockerfile logic after confirming detected services, so that it can only apply for those services confirmed by user.

File: cli/azd/internal/appdetect/docker.go [hunk 9]
```diff
@@ -70,3 +71,79 @@ func parsePortsInLine(s string) ([]Port, error) {
...
+    elif [ -f /app.war ]; then java -jar /app.war; \
+    else echo 'No JAR or WAR file found'; fi"]`
+)
+
+func writeDockerfileIntoFs(path string, hasParentPom bool) error {
+	if _, err := os.Stat(path); err != nil {
+		return fmt.Errorf("error accessing path %s: %w", path, err)
+	}
+
+	dockerfilePath := filepath.Join(path, "Dockerfile")
```
Comment (saragluna) [inline cli/azd/internal/appdetect/docker.go:121]: I have a concern here that whether we should output this Dockerfile to user's source folder. But it's okay for now, let's keep it the straightforward way before we have a better idea.

**Review by saragluna — APPROVED** [review]

File: cli/azd/internal/appdetect/docker.go (hunk 8 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go:95]: Here I will first update the base image to jdk-21. Later, we will use another item to track the Java version detection.
https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2330461

File: cli/azd/internal/appdetect/appdetect.go [hunk 10]
```diff
@@ -326,7 +326,7 @@ func detectAny(ctx context.Context, detectors []projectDetector, path string, en
 			log.Printf("Found project %s at %s", project.Language, path)
 
 			// docker is an optional property of a project, and thus is different from other detectors
-			docker, err := detectDockerInDirectory(path, entries)
+			docker, err := detectDockerInDirectory(*project, entries)
```
Comment (rujche) [inline cli/azd/internal/appdetect/appdetect.go:329]: Why is this change necessary?

File: cli/azd/internal/appdetect/docker.go (hunk 9 above)
Comment (rujche) [reply cli/azd/internal/appdetect/docker.go:122]: Here is my suggestion: 
1. Not create docker when app detect. But return the information exists when do app detect for java applications.
2. Create the dockerfile when use it to build docker image.
3. Delete the docker file after docker image build finished.

File: cli/azd/internal/repository/app_init.go [hunk 11]
```diff
@@ -464,6 +464,14 @@ func (i *Initializer) prjConfigFromDetect(
 				return config, err
 			}
 		}
+		// if java project no Dockerfile, add a default one
+		if svc.Language == appdetect.Java && svc.Docker == nil {
+			docker, err := appdetect.AddDefaultDockerfile(svc)
```
Comment (rujche) [inline cli/azd/internal/repository/app_init.go:469]: If `AddDefaultDockerfile` only used in `app_init.go`, how about moving it to current package / file?

File: cli/azd/internal/appdetect/appdetect.go (hunk 10 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/appdetect.go:329]: necessary before, but not used now, updated.

File: cli/azd/internal/repository/app_init.go (hunk 11 above)
Comment (haoozhang) [reply cli/azd/internal/repository/app_init.go:469]: This func is the part of `Docker`, and in the func body, it also refers to something of the `appdetect` package, so it would be better to place it into current `appdetect` package.

File: cli/azd/internal/appdetect/docker.go (hunk 7 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: Per offline sync with Rujun, the current detect logic is that azd java analyzer detects it as a java project as long as `pom.xml` exists, which cannot identify if it's runnable or not. We have to enhance this detect more accurately to filter out those not runnable ones. 
Let's use this item to track https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2330953

File: cli/azd/internal/repository/app_init.go (hunk 11 above)
Comment (haoozhang) [reply cli/azd/internal/repository/app_init.go]: Moved this func to `azd package` phrase.

File: cli/azd/internal/appdetect/docker.go (hunk 9 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: Per offline sync, we will move this logic into `azd package` phrase, and add the default Dockerfile into temp dir, not user source code folder.

File: cli/azd/pkg/project/framework_service_docker.go [hunk 12]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
 
 	return options
 }
+
+func addDefaultDockerfileForJava(serviceConfig ServiceConfig) (string, error) {
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:635]: `addDefaultDockerfileForJava` -> `addDefaultDockerfileForJavaProject`

File: cli/azd/pkg/project/framework_service_docker.go [hunk 13]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+	// if Dockerfile not exists, provide a default one
+	if serviceConfig.Language == ServiceLanguageJava {
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:637]: `serviceConfig.Language == ServiceLanguageJava`, this check happened twice. How about doing like this:

```go
addDefaultDockerfileForJavaProject(serviceConfig) // Pass pointer

...


func addDefaultDockerfileForJava(serviceConfig ServiceConfig) (string, error) {
  if serviceConfig.Language == ServiceLanguageJava && serviceConfig.Docker.Path == "" {...}
}
```

File: cli/azd/pkg/project/framework_service_docker.go [hunk 14]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+WORKDIR /
+COPY --from=build /app/target/*.jar app.jar
+COPY --from=build /app/target/*.war app.war
+ENTRYPOINT ["sh", "-c", \
+    "if [ -f /app.jar ]; then java -jar /app.jar; \
+    elif [ -f /app.war ]; then java -jar /app.war; \
+    else echo 'No JAR or WAR file found'; fi"]`
+)
+
+func writeDockerfileIntoFs(path string, hasParent bool) (string, error) {
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:679]: Did you test this? will this Dockerfile be used in `azd package` as expected?

File: cli/azd/pkg/project/framework_service_docker.go [hunk 15]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+
+// todo: hardcode jdk-21 as base image here, may need more accurate java version detection.
+const (
+	DockerfileSingleStage = `FROM openjdk:21-jdk-slim
+COPY ./target/*.jar app.jar
+COPY ./target/*.war app.war
+ENTRYPOINT ["sh", "-c", \
+    "if [ -f /app.jar ]; then java -jar /app.jar; \
+    elif [ -f /app.war ]; then java -jar /app.war; \
+    else echo 'No JAR or WAR file found'; fi"]`
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:662]: Is it possible to report error instead of using `echo`?

File: cli/azd/pkg/project/framework_service_docker.go [hunk 16]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+	}
+
+	return "", fmt.Errorf("adding default Dockerfile is not supported for language: %s", serviceConfig.Language)
+}
+
+// todo: hardcode jdk-21 as base image here, may need more accurate java version detection.
+const (
+	DockerfileSingleStage = `FROM openjdk:21-jdk-slim
+COPY ./target/*.jar app.jar
+COPY ./target/*.war app.war
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:658]: How about deleting `.war`, in normal case, `war` file is used to be deployed in web server. And `jar` will be run by `java -jar`.

File: cli/azd/pkg/project/framework_service_docker.go [hunk 17]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+	}
+
+	file, err := os.Create(dockerfilePath)
+	if err != nil {
+		return "", fmt.Errorf("failed to create Dockerfile at %s: %w", dockerfilePath, err)
+	}
+	defer file.Close()
+
+	// for single-module project, we have to run 'mvn package' first, then copy and run jar
+	// for multi-module project, just copy and run jar because 'mvn package' already executed in prepackage hook
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:699]: How about running `mvn package` in one place for all cases? For example: run `mvn package` for all mvn project in prepackage hook.

File: cli/azd/pkg/project/framework_service_docker.go [hunk 18]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
+		return "", fmt.Errorf("error accessing path %s: %w", path, err)
+	}
+
+	dockerfilePath := filepath.Join(path, "Dockerfile")
+	if _, err := os.Stat(dockerfilePath); err == nil {
+		fmt.Println("Dockerfile already exists, skipping creation.")
+		return dockerfilePath, nil
+	} else if !os.IsNotExist(err) {
+		return "", fmt.Errorf("error checking Dockerfile at path %s: %w", path, err)
+	}
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:690]: How about moving `create temp folder` related codes into current func, and remove all codes about file check (including check file exist and dir exists)? In new created folder, it's not necessary to check.

File: cli/azd/pkg/project/framework_service_docker.go [hunk 19]
```diff
@@ -619,3 +631,82 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
 
 	return options
 }
+
+func addDefaultDockerfileForJava(serviceConfig ServiceConfig) (string, error) {
+	// if Dockerfile not exists, provide a default one
+	if serviceConfig.Language == ServiceLanguageJava {
+		log.Printf("Dockerfile not found for java project %s, will provide a default one", serviceConfig.Name)
+		dockerfileDir, err := os.MkdirTemp("", fmt.Sprintf("%s-%d", serviceConfig.Name, time.Now().Unix()))
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:639]: Keep it align with other places:
> image

File: cli/azd/pkg/project/framework_service_docker.go (hunk 12 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:635]: Updated.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 14 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:679]: Yes, tested it. It just like the command: `docker build -t <image name> -f <docker file path> .`

File: cli/azd/pkg/project/framework_service_docker.go (hunk 15 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:662]: exit with a non-zero status code if no jar found.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 16 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:658]: yeah, we can delete it first.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 18 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:690]: simplified this logic.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 19 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:639]: Updated.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 17 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:699]: Updated.

File: cli/azd/internal/repository/app_init.go [hunk 20]
```diff
@@ -911,11 +911,18 @@ func (i *Initializer) addMavenBuildHook(
 	wrapperPathMap := map[string][]string{}
 
 	for _, prj := range detect.Services {
-		if prj.Language == appdetect.Java && prj.Options["parentPath"] != nil {
-			parentPath := prj.Options[appdetect.JavaProjectOptionMavenParentPath].(string)
-			posixMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionPosixMavenWrapperPath].(string)
-			winMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionWinMavenWrapperPath].(string)
-			wrapperPathMap[parentPath] = []string{posixMavenWrapperPath, winMavenWrapperPath}
+		if prj.Language == appdetect.Java {
+			if prj.Options["parentPath"] != nil {
```
Comment (rujche) [inline cli/azd/internal/repository/app_init.go:915]: Use const variable instead of string here. `"parentPath"` -> `appdetect.JavaProjectOptionMavenParentPath`

File: cli/azd/pkg/project/framework_service_docker.go [hunk 21]
```diff
@@ -32,6 +26,11 @@ import (
 	"github.com/azure/azure-dev/cli/azd/pkg/tools/docker"
 	"github.com/azure/azure-dev/cli/azd/pkg/tools/pack"
 	"go.opentelemetry.io/otel/trace"
+	"log"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:33]: revert this change. Don't add unnecessary changes.

You can configure your Go Land like this:

> image

File: cli/azd/pkg/project/framework_service_docker.go [hunk 22]
```diff
@@ -619,3 +630,29 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
 
 	return options
 }
+
+// todo: hardcode jdk-21 as base image here, may need more accurate java version detection.
+const DefaultDockerfileForJavaProject = `FROM openjdk:21-jdk-slim
+COPY ./target/*.jar app.jar
+ENTRYPOINT ["sh", "-c", \
+    "if [ -f /app.jar ]; then java -jar /app.jar; \
+    else echo 'Error: No JAR file found' >&2; exit 1; fi"]`
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:639]: Is it possible to report this error at build time instead of runtime?

File: cli/azd/internal/repository/app_init.go [hunk 23]
```diff
@@ -911,11 +911,18 @@ func (i *Initializer) addMavenBuildHook(
...
+				parentPath := prj.Options[appdetect.JavaProjectOptionMavenParentPath].(string)
+				posixMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionPosixMavenWrapperPath].(string)
+				winMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionWinMavenWrapperPath].(string)
+				wrapperPathMap[parentPath] = []string{posixMavenWrapperPath, winMavenWrapperPath}
+			} else {
+				prjPath := prj.Options[appdetect.JavaProjectOptionMavenPath].(string)
+				posixMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionPosixMavenWrapperPath].(string)
+				winMavenWrapperPath := prj.Options[appdetect.JavaProjectOptionWinMavenWrapperPath].(string)
+				wrapperPathMap[prjPath] = []string{posixMavenWrapperPath, winMavenWrapperPath}
```
Comment (rujche) [inline cli/azd/internal/repository/app_init.go:924]: After this PR merged, you can consider this: What if `java` not exist in customer's environment? In that case, `mvn package` will return error. Of course, you can create a ADO item to track this.

File: cli/azd/internal/appdetect/appdetect_test.go [hunk 24]
```diff
@@ -307,6 +322,11 @@ func TestDetect(t *testing.T) {
 					Language:      Java,
 					Path:          "java",
 					DetectionRule: "Inferred by presence of: pom.xml",
+					Options: map[string]interface{}{
+						JavaProjectOptionMavenPath:             filepath.Join(dir, "java"),
+						JavaProjectOptionPosixMavenWrapperPath: filepath.Join(dir, "java", "mvnw"),
+						JavaProjectOptionWinMavenWrapperPath:   filepath.Join(dir, "java", "mvnw.cmd"),
```
Comment (rujche) [inline cli/azd/internal/appdetect/appdetect_test.go:328]: Could you please help to improve this (mvn command related codes) after this PR merged: https://github.com/azure-javaee/azure-dev/pull/100

File: cli/azd/internal/appdetect/java.go [hunk 25]
```diff
@@ -21,6 +21,9 @@ type mavenWrapper struct {
 	winPath   string
 }
 
+// JavaProjectOptionMavenPath The project path of the maven single-module project
+const JavaProjectOptionMavenPath = "path"
```
Comment (rujche) [inline cli/azd/internal/appdetect/java.go:25]: Nit: rename to `JavaProjectOptionCurrentPomDir` and `JavaProjectOptionParentPomDir` will make it easier to understand. `JavaProjectOptionMavenPath` is confusing without read the codes where used it.

File: cli/azd/internal/appdetect/appdetect_test.go (hunk 24 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/appdetect_test.go:328]: Ok, will handle this. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2331044

File: cli/azd/internal/repository/app_init.go (hunk 20 above)
Comment (haoozhang) [reply cli/azd/internal/repository/app_init.go]: Updated.

File: cli/azd/internal/appdetect/java.go (hunk 25 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/java.go]: Updated.

File: cli/azd/internal/repository/app_init.go (hunk 23 above)
Comment (haoozhang) [reply cli/azd/internal/repository/app_init.go:924]: Will use this item to track first. https://devdiv.visualstudio.com/DevDiv/_workitems/edit/2331052
Then we can have further discussion for how to deal with this kind of case.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 21 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go]: Updated, thanks!

File: cli/azd/pkg/project/framework_service_docker.go (hunk 22 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:641]: It's in the bash script, normally it returns non-zero code to indicate any error. I don't have any other better idea for now. 

BTW, I think we can have the assumption that after successfully `mvn clean package` (we added the hook), the output jar should be located at the `target` folder, so even we don't need the check in the line 639, just `java -jar *.jar`.

File: cli/azd/internal/appdetect/appdetect_test.go (hunk 24 above)
Comment (rujche) [reply cli/azd/internal/appdetect/appdetect_test.go:328]: Thank you. And maybe support multiple pom files in one folder in the new task. For example:
```
./project
./project/frontend-pom.xml
./project/backend-pom.xml
```

In this case, maybe we can change `JavaProjectOptionCurrentPomDir` to `JavaProjectOptionCurrentPomPath`

File: cli/azd/pkg/project/framework_service_docker.go [hunk 26]
```diff
@@ -200,6 +201,18 @@ func (p *dockerProject) Build(
 		return &ServiceBuildResult{Restore: restoreOutput}, nil
 	}
 
+	// if java project no Dockerfile, add a default one for Docker build
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:204]: nit: `if java project no Dockerfile` -> `if it's a java project without Dockerfile`

File: cli/azd/pkg/project/framework_service_docker.go (hunk 22 above)
Comment (rujche) [reply cli/azd/pkg/project/framework_service_docker.go:641]: > so even we don't need the check in the line 639, just java -jar *.jar is enough?

Agree, we can change to this:
```shell
ENTRYPOINT ["sh", "-c", "java -jar /app.jar"]
```

BTW, please confirm after using base image `openjdk:21-jdk-slim`, the working directory is root dir `/`. So after `COPY ./target/*.jar app.jar`, the `app.jar` will be put in root dir. I suggest to change `COPY ./target/*.jar app.jar` to `COPY ./target/*.jar /app.jar` to make our code more robust.

Comment (rujche) [reply cli/azd/pkg/project/framework_service_docker.go:641]: > It's in the bash script, normally it returns non-zero code to indicate any error. I don't have any other better idea for now. What do you think about it?

If `./target/*.jar` not exist, will `COPY ./target/*.jar app.jar` make `docker build` failed?

File: cli/azd/pkg/project/framework_service_docker.go [hunk 27]
```diff
@@ -619,3 +632,29 @@ func getDockerOptionsWithDefaults(options DockerProjectOptions) DockerProjectOpt
...
 }
+
+// todo: hardcode jdk-21 as base image here, may need more accurate java version detection.
+const DefaultDockerfileForJavaProject = `FROM openjdk:21-jdk-slim
+COPY ./target/*.jar app.jar
+ENTRYPOINT ["sh", "-c", \
+    "if [ -f /app.jar ]; then java -jar /app.jar; \
+    else echo 'Error: No JAR file found' >&2; exit 1; fi"]`
+
+func addDefaultDockerfileForJavaProject(svcName string) (string, error) {
```
Comment (rujche) [inline cli/azd/pkg/project/framework_service_docker.go:643]: BTW, what's the merit after using default Dockerfile comparing to `build from source`? IMU, without Dockerfile, `build from source` will be used.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 22 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:641]: > change `COPY ./target/*.jar app.jar` to `COPY ./target/*.jar /app.jar` to make our code more robust.

The default working dir is `/`, and also updated the COPY command.

> If `./target/*.jar` not exist, will `COPY ./target/*.jar app.jar` make docker build failed?

Test from my side, it will throw an error:
image

File: cli/azd/internal/appdetect/appdetect_test.go (hunk 24 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/appdetect_test.go:328]: I currently have no clear context for the PR https://github.com/azure-javaee/azure-dev/pull/100, will consider your mentioned case, thanks for your reminder.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 27 above)
Comment (haoozhang) [reply cli/azd/pkg/project/framework_service_docker.go:643]: During internal bug bash, some feedback indicate that the pack build (build from source) may stuck or cost too much time, so we consider adding this default Dockerfile to build from Dockerfile and work around too-slow pack build, potentially to save some time.

File: cli/azd/internal/appdetect/docker.go (hunk 6 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: PR merged https://github.com/azure-javaee/azure-dev/pull/94.

File: cli/azd/pkg/project/framework_service_docker.go (hunk 27 above)
Comment (rujche) [reply cli/azd/pkg/project/framework_service_docker.go:641]: Got it. Maybe we should back to `build from source`  if `java` command not exist in customer's path.

**Review by rujche — APPROVED** [review]
LGTM.

File: cli/azd/internal/appdetect/docker.go (hunk 4 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: See this fix PR: https://github.com/azure-javaee/azure-dev/pull/103

File: cli/azd/internal/appdetect/docker.go (hunk 7 above)
Comment (haoozhang) [reply cli/azd/internal/appdetect/docker.go]: See this fix PR: https://github.com/azure-javaee/azure-dev/pull/103
